### PR TITLE
Add support of Amazon EFA

### PR DESCRIPTION
**Pull request by DwarKapex**
Address issue: EFA Support #167

    The upstream JAX container contains only MOFED NIC support. The MOFED package
    from Mellanox that we use installs *ibverbs* libraries which do not contain
    libefa*.so which are required for AWS.
    
    A temporary solution is to provide a script as a part of base container
    (/usr/local/bin/install-efa.sh) that AWS user can run inside the container
    to handle this issue.
    The script does the following:
    1. Remove all *ibverbs* and RDMA related libraries
    2. Download Amazon EFA installer
    3. Install EFA

**How to use**: in the running container run **install-efa.sh** script:
`
root@<container-id> $> install-efa.sh
`

### CODE REVIEW

File: .github/container/install-efa.sh
Comment (yhtang) [inline .github/container/install-efa.sh:35]: Environment variables must be set using the `ENV` statement in the Dockerfile to persist.

File: .github/container/install-efa.sh
Comment (yhtang) [inline .github/container/install-efa.sh:16]: Are the existing ibverbs packages installed by the `install-ofed.sh` script?

**Review by yhtang — CHANGES_REQUESTED** [review]

File: .github/container/install-efa.sh
Comment (yhtang) [reply .github/container/install-efa.sh:16]: Per offline communication, the answer is yes, and we need to sort out a way to reconcile MOFED and EFA.

Comment (DwarKapex) [reply .github/container/install-efa.sh]: Current solution is to install required libs from Ubuntu upstream apt repository explicitly.

**Comment by sbhavani** [comment]
Do we have support for all the packages required to run on AWS? Here's a reference for customizing an NGC image: https://github.com/aws-samples/aws-efa-nccl-baseami-pipeline/blob/master/nvidia-efa-docker_base/ubuntu18.04/Dockerfile-cu11-pt20.10.sagemaker

File: .github/container/install-efa.sh
Comment (mjsML) [reply .github/container/install-efa.sh]: Are they different or just a dependency conflict ?

**Comment by DwarKapex** [comment]
> Do we have support for all the packages required to run on AWS? Here's a reference for customizing an NGC image: https://github.com/aws-samples/aws-efa-nccl-baseami-pipeline/blob/master/nvidia-efa-docker_base/ubuntu18.04/Dockerfile-cu11-pt20.10.sagemaker

Not sure, need to test. NGC image that you suggested uses efa-installer from amazon. That was my initial approach, but I found a conflict between efa-installer and MOFED: both installs **libibverbs** package. Interesting, that MOFED implementation of **libibverbs** does not contain libefa*, meaning, that if the installation order is MOFED -> efa-installer, MOFED implementation will be overwritten, in other way around efa-instller -> MOFED, libefa* will be unavailable.

@yhtang what is so important in MOFED? Can we use Amazon implementation instead?

File: .github/container/install-efa.sh
Comment (DwarKapex) [reply .github/container/install-efa.sh]: They are different

File: .github/container/install-efa.sh
Comment (yhtang) [inline .github/container/install-efa.sh:32]: This would make the execution outcome of install-efa/ofed.sh order-dependent, and this order is not explicit.
Why do we have to remove the two packages if they are already installed?

**Review by yhtang — CHANGES_REQUESTED** [review]

File: .github/container/install-efa.sh
Comment (yhtang) [reply .github/container/install-efa.sh:32]: Could we at least raise a warning if the two packages are found pre-existing in the system? Or maybe add an argument `--remove-existing-ibverbs` using shell script options?

**Comment by sbhavani** [comment]
I think we should have a separate image or static Dockerfile example for AWS. 

I remember seeing that some of our proprietary MOFED bits conflict with Amazon's so it's been difficult to ship both in NGC. 

+ @mjsML any thoughts?

**Comment by mjsML** [comment]
sgtm

**Comment by sbhavani** [comment]
I don't think we need to call out IB, it's included by default in NGC containers and doesn't need a separate tag. Could we keep the extra tag just for efa?

**Comment by andportnoy** [comment]
> I think we should have a separate image or static Dockerfile example for AWS.

> nitpik, it would be on the jax container, so it will propagate to all containers: {jax,t5x,pax}.IB.py3 and {jax,t5x,pax}.EFA.py3

Use [multi-stage builds](https://docs.docker.com/build/building/multi-stage/) like we do already for the JAX container?

Then we could build two separate images using the same Dockerfile:
```bash
docker build -f Dockerfile.base --target generic --tag base-generic .
docker build -f Dockerfile.base --target aws     --tag base-aws     .
```

**Comment by DwarKapex** [comment]
Make changes per today's morning discussions:         
As a workaround the strategy is the following:
1. Build the base image with MOFED support installed.
2. Add an extra script to the image (/usr/local/bin/install-efa.sh)  to be able to install Amazon EFA per request.

@yhtang **module** cmd is an interesting approach, but IMHO in this case is overkill. How to add a release note on **How to use it**?

File: .github/container/install-efa.sh
Comment (DwarKapex) [inline .github/container/install-efa.sh:28]: Add full spec of flags: `-y -g -d --skip-kmod --skip-limit-conf --no-verify` @DwarKapex

**Comment by sbhavani** [comment]
> Make changes per today's morning discussions: As a workaround the strategy is the following:
> 
> 1. Build the base image with MOFED support installed.
> 2. Add an extra script to the image (/usr/local/bin/install-efa.sh)  to be able to install Amazon EFA per request.
> 
> @yhtang **module** cmd is an interesting approach, but IMHO in this case is overkill. How to add a release note on **How to use it**?

for the how to use it, we should try to keep the library versions as close to what AWS supports for it's DL containers: https://github.com/aws/deep-learning-containers/blob/master/tensorflow/training/docker/2.13/py3/cu118/Dockerfile.gpu

**Comment by DwarKapex** [comment]
> for the how to use it, we should try to keep the library versions as close to what AWS supports for it's DL containers: https://github.com/aws/deep-learning-containers/blob/master/tensorflow/training/docker/2.13/py3/cu118/Dockerfile.gpu

I think, they manually increment the version of Amazon EFA installer. Don't know if it is a good idea, but if it's really matter I can keep in sync the DL containers version and ours, not a big deal (@sbhavani what do you think?). As a reference: they use version 1.24.0, we use `latest`, which is 1.25.0 as of now.

**Comment by yhtang** [comment]
This is at least good for unblocking people that need EFA, so I'd say let's go ahead and merge it.

In the long run, I still believe a more systematic approach, such as the `modulefiles` solution is better in terms of unifying the container bases. There are potential drawbacks of letting users run the installer script after pulling the image, including
- making the container no longer self-contained for archival and bug-repro purposes.
- requires network connection when running the container (may be mitigated by distributing the downloaded installer package, but bloats image)
- longer pull time, bigger image.
- etc...

**Review by yhtang — DISMISSED** [review]

**Comment by yhtang** [comment]
Before merging, @DwarKapex could you please update the PR description so that it documents the approach taken, and that it is a stopgap solution to unblock users, while further work is still needed to harden it?

**Review by yhtang — APPROVED** [review]

**Comment by yhtang** [comment]
/assistant summarize the key takeaways of the discussion using concise bullet points

**Pull request merged**
